src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp

```diff
@@ -82,11 +82,17 @@ inline oop ShenandoahForwarding::try_update_forwardee(oop obj, oop update) {
   }
 
   markWord new_mark = markWord::encode_pointer_as_mark(update);
-  markWord prev_mark = obj->cas_set_mark(new_mark, old_mark, memory_order_conservative);
-  if (prev_mark == old_mark) {
-    return update;
-  } else {
-    return cast_to_oop(prev_mark.clear_lock_bits().to_pointer());
+  while (true) {
+    markWord prev_mark = obj->cas_set_mark(new_mark, old_mark, memory_order_conservative);
+    if (prev_mark == old_mark) {
+      return update;
+    } else if (prev_mark == markWord::INFLATING()) {
+      // This happens when we encounter a stack-locked object in from-space.
+      // Busy-wait for completion.
```

Comment:
Can you explain how this can happen? I think we can only inflate to-space object, e.g. apply LRB first.

Reply:
The trouble is that we need to fetch the size of an object during evacuation *before* the object is guaranteed to be in to-space. If we encounter an object that is stack-locked there, we may race with another thread which could, in the meantime, succeed to evacuate and unlock the object, in which case the stack-lock that we fetched from the from-space object would be invalid. We prevent that by CASing 0 into the header (e.g. the INFLATING marker) - this prevents both evacuation and unlocking, but it requires that our evacuation CAS is aware of it and retries if that happens.

Reply:
Hmm... it hurts my head to think about if we can see this transition state in from-space object, e.g. `ShenandoahForwarding::is_forwarded()` can return bogus result? all the from-space reads have to be stable reads?

Reply:
is_forwarded() will always be good: 00 means stack-locked or inflating, 01 means unlocked, 10 means monitor and 11 means forwarded.
There are places where from-space header accesses (to get object size and/or Klass*) requires extra care: all those cases have been changed to use the new ShenandoahObjectUtils functions instead of the usual stuff, and this helper takes care of dealing with concurrent evacuation (i.e. the otherwise unexpected 11 state).

Reply:
Okay, let's consider following scenario:
1. worker 1 tries to evac obj, reads its header and it is not `marked`
2. worker 1 reads obj size and allocates to-space obj
3. worker 2 tried to evac the same obj, reads its header and it is still not `marked`
4. worker 2 tries to read obj size and temporary installs `INFLATING` mark word
5. worker 1 copies obj to to-space with `INFLATING` mark word (?)
6. worker 1 wins CAS, so to-space obj has `INFLATING' mark word?

Reply:
At 6, worker 1 cannot win while the from-space object has INFLATING in its header. It will always try to CAS with the real original header. That is why we need the loop in the code that installs the forwarding pointer to retry when encountering INFLATING.

Reply:
Line #[79](https://github.com/openjdk/lilliput/blob/master/src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp#L79) is not a stable read, right? so it can see `INFLATING` markword. I don't see why CAS at Line#[85](https://github.com/openjdk/lilliput/blob/master/src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp#L85) has to fail? it may return `prev_mark = INFLATING`, no?